storage_controller/src/tenant_shard.rs

```diff
@@ -1351,6 +1352,32 @@ impl TenantShard {
     pub(crate) fn set_preferred_az(&mut self, preferred_az_id: AvailabilityZone) {
         self.preferred_az_id = Some(preferred_az_id);
     }
+
+    pub(crate) fn attached_locations(&self) -> Vec<(NodeId, Generation)> {
+        self.observed
+            .locations
+            .iter()
+            .filter_map(|(node_id, observed)| {
+                let conf = observed.conf.as_ref()?;
+                let is_attached = matches!(
+                    conf.mode,
+                    LocationConfigMode::AttachedMulti
+                        | LocationConfigMode::AttachedSingle
+                        | LocationConfigMode::AttachedStale
+                );
+                let has_generation = conf.generation.is_some();
+                if is_attached && has_generation {
+                    Some((*node_id, conf.generation.unwrap()))
+                } else {
+                    None
+                }
+            })
+            .sorted_by(|(_lhs_node_id, lhs_gen), (_rhs_node_id, rhs_gen)| {
+                lhs_gen.cmp(rhs_gen).reverse()
+            })
+            .map(|(node_id, gen)| (node_id, Generation::new(gen)))
+            .collect()
```

Comment:
FYI with [`sorted_by`](https://docs.rs/itertools/latest/src/itertools/lib.rs.html#2906-2914) this will create and allocate two vecs (one at sorted_by, other at collect).

Reply:
I guess I could sort after collect, but the size of the max size of this vector should always be 2 or 3, so I don't think it matters that much.